metric_metadata/cached/cached.go

```diff
@@ -15,68 +15,118 @@
+	TimeToLive   time.Duration
+}
+
+// CachedTagSetList is an item in the cache.
+type CachedTagSetList struct {
+	TagSets []api.TagSet // The tagsets for this metric
+	Expiry  time.Time    // The time at which the cache entry expires
+	Stale   time.Time    // The time at which the cache entry becomes stale
+
+	sync.Mutex // Synchronizing mutex
+
+	inflight bool           // Indicates a request is already in flight
+	enqueued bool           // Indicates a request has been enqueued
+	wg       sync.WaitGroup // Synchronizing wait group
+
+	fetchError error // Fetch error from the last attempt
 }
 
 // NewMetricMetadataAPI creates a cached API given configuration and an underlying API object.
 func NewMetricMetadataAPI(apiInstance metadata.MetricAPI, config Config) *CachedMetricMetadataAPI {
 	requests := make(chan func(metadata.Context) error, config.RequestLimit)
+	if config.Freshness == 0 {
+		config.Freshness = config.TimeToLive
+	}
 	return &CachedMetricMetadataAPI{
 		metricMetadataAPI: apiInstance,
-		getAllTagsCache:   map[api.MetricKey]CachedTagSetList{},
+		clock:             util.RealClock{},
+		getAllTagsCache:   map[api.MetricKey]*CachedTagSetList{},
+		freshness:         config.Freshness,
 		timeToLive:        config.TimeToLive,
 		backgroundQueue:   requests,
 	}
 }
 
-// CachedTagSetList is an item in the cache.
-type CachedTagSetList struct {
-	TagSets []api.TagSet // The tagsets for this metric
-	Expiry  time.Time    // The time at which the cache entry expires
-}
+// addBackgroundGetAllTagsRequest adds a job to update the lag list for the given
+// metric. Requires the caller hold the lock for the item in the cache.
+func (c *CachedMetricMetadataAPI) addBackgroundGetAllTagsRequest(metricKey api.MetricKey) {
+	c.queueMutex.Lock()
+	defer c.queueMutex.Unlock()
 
-// Expired tells whether the entry has zero-time (meaning absent) or is out-of-date.
-func (c CachedTagSetList) Expired() bool {
-	return c.Expiry.IsZero() || c.Expiry.Before(time.Now())
-}
+	if cap(c.backgroundQueue) <= len(c.backgroundQueue) {
+		log.Warningf("Unable to enqueue a background GetAllTags lookup for %s due to a full queue", metricKey)
+		return
+	}
 
-func (c *CachedMetricMetadataAPI) addBackgroundGetAllTagsRequest(metric api.MetricKey) {
-	c.mutex.Lock()
-	defer c.mutex.Unlock()
-	if len(c.backgroundQueue) < cap(c.backgroundQueue) {
-		// Add this function to the background queue.
-		c.backgroundQueue <- func(context metadata.Context) error {
-			defer context.Profiler.Record("CachedMetricMetadataAPI_BackgroundAction_GetAllTags")()
-			startTime := time.Now()
-			tagsets, err := c.metricMetadataAPI.GetAllTags(metric, context)
-			if err != nil {
-				return err
-			}
-			c.replaceCachedTagSet(metric, tagsets, startTime)
-			return nil
-		}
+	// No need to lock on getAllTagsCache since we already should hold the lock for
+	// the given item, so it should have already been written in a thread-safe way
+	item, ok := c.getAllTagsCache[metricKey]
```

Comment:
There's still nothing protecting this read:

Just because we have a lock on `item` doesn't mean that someone isn't modifying a different metricKey in the cache: this causes a race condition.

We either need to acquire the `getAllTagsCacheMutex` before this read, or we need the caller to do it.

Since in all cases that this function is called, the caller already retrieved `item` from the map themselves a few lines, it seems simpler to me to just ask for `item` as a parameter and avoid the map lookup completely.